R/sagemaker.R

```diff
@@ -0,0 +1,481 @@
+#' that has been versioned and stored via [vetiver_pin_write()] as a Plumber API
+#' on Amazon SageMaker.
+#'
+#' @inheritParams vetiver_sm_build
+#' @inheritParams vetiver_sm_model
+#' @inheritParams vetiver_sm_endpoint
+#' @param repo_name The name for the AWS ECR repository to store the model.
+#' @param ... Other arguments passed to [vetiver_sm_endpoint()] such as
+#' `accelerator_type` or `data_capture_config`.
+#'
+#' @details
+#' This function stores your model deployment image in the same bucket used
+#' by `board`.
+#'
+#' The function `vetiver_deploy_sagemaker()` uses:
+#' - [vetiver_sm_build()] to build and push a Docker image to ECR,
+#' - [vetiver_sm_model()] to create a SageMaker model, and
+#' - [vetiver_sm_endpoint()] to deploy a SageMaker model endpoint.
+#'
+#' These modular functions are available for more advanced use cases.
+#'
+#' @return
+#' The deployed SageMaker model endpoint.
+#'
+#' @seealso [vetiver_sm_build()], [vetiver_sm_model()], [vetiver_sm_endpoint()]
+#' @export
+#' @examples
+#' if (FALSE) {
+#' library(pins)
+#' b <- board_s3(bucket = "my-existing-bucket")
+#' cars_lm <- lm(mpg ~ ., data = mtcars)
+#' v <- vetiver_model(cars_lm, "cars_linear")
+#' vetiver_pin_write(b, v)
+#'
+#' endpoint <- vetiver_deploy_sagemaker(
+#'     board = b,
+#'     name = "cars_linear",
+#'     predict_args = list(type = "class", debug = TRUE),
+#'     instance_type = "ml.t2.medium"
+#' )
+#' }
+#'
+vetiver_deploy_sagemaker <- function(board,
+                                     name,
+                                     version = NULL,
+                                     predict_args = list(),
+                                     docker_args = list(),
+                                     repo_name = glue("vetiver-sagemaker-{name}"),
+                                     compute_type = c(
+                                         "BUILD_GENERAL1_SMALL",
+                                         "BUILD_GENERAL1_MEDIUM",
+                                         "BUILD_GENERAL1_LARGE",
+                                         "BUILD_GENERAL1_2XLARGE"
+                                     ),
+                                     instance_type = "ml.t2.medium",
+                                     ...) {
+
+    if (!inherits(board, "pins_board_s3")) {
+        stop_input_type(board, "an S3 pins board")
+    }
+
+    repo_name <- ifelse(
+        grepl(":", repo_name),
+        repo_name,
+        glue("{repo_name}:{strftime(Sys.time(), '%Y-%m-%d')}")
+    )
+
+    # build image and push to aws ecr
+    image_uri <- vetiver_sm_build(
+        board = board,
+        name = name,
+        version = version,
+        predict_args = predict_args,
+        docker_args = docker_args,
+        repository = repo_name,
+        compute_type = compute_type,
+        bucket = board$bucket
+    )
+
+    args <- list2(...)
+    tags <- list_modify(args$tags, "vetiver")
```

Comment:
> This is looking pretty good! There is still some work to be done around the tags for the model and endpoint.

I believe the issues is down to this that. It should be in the following format

```r
list(
    list(
        'Key'='string',
        'Value'='string'
     )
)
```
This is my bad for not documenting it correctly, sorry about that

Reply:
Got an initial idea for making tagging easier, I will push it up so i can do an initial test on my sagemaker environment.